### DISCUSSION TITLE
What's the best way to use the Pi Debug Probe with this library?

### DISCUSSION THREAD

**Discussion by Zelzahn** [tinygo-org/tinygo#4240] [Q&A]
Greetings, I'm coming from Rust so Go + Embedded is still a bit new for me. I have a Raspberry Pi Pico linked up with a Pi Debug Probe. Currently, I do `tinygo build -target=pico main.go`, followed with `probe-rs run --chip RP2040 main.elf`.

Is there any way to make `tinygo flash` call the `probe-rs` command via any config file? Or is there any other tooling better integrated?

**Comment by kenbell** [comment]
I haven't played with it for a couple years, but with a suitable version of openocd you should be able to do things like:
`tinygo gdb -target=pico main.go`

Last time i played with it, you needed a custom / non-released version of openocd - but it looks like the 0.12.0 release should have everything needed: https://openocd.org/openocd-0120-released.html

**Comment by dufguix** [comment]
Why do you use probe-rs ?
I am curious.

**Comment by aykevl** [comment]
Usually `tinygo flash -target=pico` will work out of the box. I _think_ you can also use `prope-rs run` but I haven't tried it.

(I think `probe-rs` is a really clean debug probe tool, I might investigate using it for TinyGo instead of OpenOCD at some point).

**Comment by lgamma** [comment]
Hi there, I'm moving the first steps with TinyGo in order to evaluate performance and development effort for a new project based on RP2040 (or maybe 2350). Looks very easy to manage source code and tool-chain.

What is pedantic, in my opinion, especially on first usage is compiling as UF2, press BOOT/RUN and flash the image to the target.

I'm used to compile directly to a binary image (evt. HEX format) then upload via **SWD** and have an UART monitor always connected to the PC instead of going through the USB device.

According to the documentation, it is possible to specify a .bin extension to the -o option in order to generate a binary image.

So I just got the basic [blinky](https://tinygo.org/docs/tutorials/blinky/) example to compile then upload with two simple bash scripts.

Build script:
```
#!/bin/bash

TARGET=pico
IMAGE=hello-pico.bin

tinygo build \
-target=$TARGET \
-serial=none \
-scheduler=none \
-panic=trap \
-o=$IMAGE \
-size=full
```

Upload script:
```
#!/bin/bash

SERIAL=E6616407E35E482F # RASPBERRY PI debug probe (original)

TARGET=rp2040
IMAGE=hello-pico.bin

edbg \
-b \
-s $SERIAL \
-t $TARGET \
-x 100 \
-c 16000 \
-epv \
-f $IMAGE -o 0x0
```

Loading generated .bin with a simple [EDBG](https://github.com/ataradov/edbg) (CLI tool) the target device remains mute even after successful programming and verification.

Here the full output for reference:

```
(.venv) luca@desktop2:~/tinygo/hello-pico$ ./build.sh 
   code  rodata    data     bss |   flash     ram | package
------------------------------- | --------------- | -------
      0       6       0       2 |       6       2 | (padding)
    372       0       0       7 |     372       7 | (unknown)
    240       0       0       0 |     240       0 | /usr/local/lib/tinygo/targets
    866       0       0       0 |     866       0 | C compiler-rt-builtins
      0     168       0       0 |     168       0 | C interrupt vector
     14       0       0       0 |      14       0 | C picolibc
      0       0       0    4096 |       0    4096 | C stack
     20       0       0       0 |      20       0 | device/arm
     18       0       0       0 |      18       0 | device/rp
    316       0       0      41 |     316      41 | machine
     44       0       0       0 |      44       0 | main
    664       0       0      38 |     664      38 | runtime
    180       0       0       0 |     180       0 | runtime/volatile
     28       0       0       0 |      28       0 | sync/atomic
------------------------------- | --------------- | -------
   2762     174       0    4184 |    2936    4184 | total

(.venv) luca@desktop2:~/tinygo/hello-pico$ ./upload.sh 
Debugger: Raspberry Pi Debug Probe (CMSIS-DAP) E6616407E35E482F 2.0.0 (S, 2)
Using CMSIS-DAP v2
Clock frequency: 16.0 MHz
Resetting... done.
Target: RP2040 (Rev B2)
Flash size: 2 MB
Erasing......... done.
Programming... done.
Verification.... done.
```

Loading the bin image to Hex Editor (VS Code extension), this file appears as 12 consecutives UF2 header blocks of 512 bytes.

May be the .bin export broken for "pico" target?

**Reply by aykevl** [reply]
You can use SWD instead if you prefer, by connecting a debug probe and adding a flag like `-programmer=cmsis-dap` or `-programmer=stlink` to TinyGo. It will use the specified programmer with OpenOCD instead.
